### PR TITLE
No Content-Type text/html on empty exception views [4.x]

### PR DESCRIPTION
**Pull request by mauritsvanrees**
Only set response header Content-Type as text/html on exception views when the response has content.

This fixes a problem in Plone with caching and 304 NotModified.
I need to describe this more, and fix the tests. Need to go now though.

### CODE REVIEW

**Comment by fredvd** [comment]
https://community.plone.org/t/warning-possibility-of-plone-5-2-10-2-sites-breaking-with-varnish-cache-and-plone-recipe-varnish/16112

**Comment by mauritsvanrees** [comment]
This is ready for review. I describe the underlying problem in issue #1089.

BTW, the gh-actions checks are not running, I am not sure why. Was this somehow switched off for PRs targeting Zope 4 because it is officially no longer supported?

I see `tests / ${{ matrix.os[0] }}-${{ matrix.config[1] }} (pull_request) Skipped`, but I think this is reported on every PR. I never understand this line.

I do hope we can still get a new Zope 4.x release, as this fixes a regression introduced by a recent release.

**Comment by dataflake** [comment]
> BTW, the gh-actions checks are not running, I am not sure why. Was this somehow switched off for PRs targeting Zope 4 because it is officially no longer supported?

No, there was nothing switched off anywhere.


> I see `tests / ${{ matrix.os[0] }}-${{ matrix.config[1] }} (pull_request) Skipped`, but I think this is reported on every PR. I never understand this line.

The check was introduced a long time ago to prevent the whole battery of tests running *twice* on every push to a PR, because two events (`push` and `pull_request`) are matched. I do not know why it is failing here, it still works on `master` and on PRs to `master`. I won't invest time in that, though, if the tests run through locally and you merge then you will always get a test run after the merge right on the `4.x` branch.


> I do hope we can still get a new Zope 4.x release, as this fixes a regression introduced by a recent release.

I can still do them, but would like someone else to take over doing these releases. My interest in Zope 4 has not just fallen to zero, it has fallen below zero. I personally consider it an unnecessary time drain for me.

File: CHANGES.rst
Comment (dataflake) [inline CHANGES.rst:13]:
```suggestion
- Only set response header Content-Type as text/html on exception views when
  the response has content.
  (`#1089 <https://github.com/zopefoundation/Zope/issues/1089>`_)
```

File: src/ZPublisher/WSGIPublisher.py
Comment (dataflake) [inline src/ZPublisher/WSGIPublisher.py:155]: This is not Plone or Varnish-specific.

```suggestion
        # An empty body may indicate a 304 NotModified response,
        # and setting a content type header will change the stored header
        # in caching servers such as Varnish.
```

File: src/ZPublisher/WSGIPublisher.py
Comment (dataflake) [inline src/ZPublisher/WSGIPublisher.py:156]: Refer to the issue and not the PR

```suggestion
        # See https://github.com/zopefoundation/Zope/issues/1089
```

**Review by dataflake — CHANGES_REQUESTED** [review]

**Review by dataflake — APPROVED** [review]

**Comment by mauritsvanrees** [comment]
Merged. Thanks for the review!

**Comment by mauritsvanrees** [comment]
Well, no, I can't actually merge. Let me try manually on the command line.

**Comment by mauritsvanrees** [comment]
Actually, I can't push manually:

```
$ git push
Enumerating objects: 14, done.
Counting objects: 100% (14/14), done.
Delta compression using up to 16 threads
Compressing objects: 100% (8/8), done.
Writing objects: 100% (8/8), 8.33 KiB | 4.16 MiB/s, done.
Total 8 (delta 6), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (6/6), completed with 6 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/4.x.
remote: error: At least 1 approving review is required by reviewers with write access. 6 of 6 required status checks are expected.
To github.com:zopefoundation/Zope.git
 ! [remote rejected]     4.x -> 4.x (protected branch hook declined)
error: failed to push some refs to 'github.com:zopefoundation/Zope.git'
```

Maybe someone else can?

BTW, the tests pass locally, except for the lint tests. And `tox -e isort-apply` does not fix it. I was hoping it was a local-only problem, with isort config being picked up from a higher directory. I have fixed it now with extra commits on the PR branch.

**Comment by dataflake** [comment]
I have removed the checks for the `4.x` branch, please try again

**Comment by mauritsvanrees** [comment]
> I can still do them, but would like someone else to take over doing these releases. My interest in Zope 4 has not just fallen to zero, it has fallen below zero. I personally consider it an unnecessary time drain for me.

I understand your point. Thank you for doing the release anyway!

Is anyone else willing to step up for doing possible Zope 4.x releases?
@icemac, are you still using Zope 4 in projects?

I know, it is officially out of support already, so new releases should be an exception in case of serious regressions or security problems.
In a pinch I could probably step up, as Plone Release Manager, but I have too many packages that I release already, with too many branches...

**Comment by icemac** [comment]
@mauritsvanrees Besides some migration project I do not use Zope 4 any more: everything else is already running on Zope 5. I could cut a Zope 4 release. Is everything on the branch needed for the release or are there still some things open?

**Comment by dataflake** [comment]
@icemac the release is done already.

**Comment by icemac** [comment]
@dataflake Thank you for cutting the release.

**Pull request merged**
